### PR TITLE
Add CMAKE option for optimizing performance of benchmarks

### PR DESCRIPTION
**Pull request by jessica-mitchell**
This PR adds the cmake option discussed with the developers about performance optimization.
A link is also added from the benchmarking page.

@heplesser @ackurth @otcathatsya  Please take a look if this is reasonable. I was not sure where in the list of cmake options it would best fit, and if the language used is fully appropriate.

### CODE REVIEW

File: doc/htmldoc/installation/cmake_options.rst [hunk 1]
```diff
@@ -99,6 +99,28 @@ Use Python to build PyNEST
...
+|                                          | for the specific type of CPU that you're      |
+|                                          | using to compile NEST                         |
++------------------------------------------+-----------------------------------------------+
+| ``-Dwith-defines=-DNDEBUG``              | disables debugging of code                    |
++------------------------------------------+-----------------------------------------------+
+
+.. note::
+
+   These options should *only be used after* you have performed several simulations of a given model with the
+   ``-Dwith-optimize="-O2"``  option.
```
Comment (heplesser) [inline doc/htmldoc/installation/cmake_options.rst:120]: I would not tie this to benchmarking so closely, it is also relevant for scientific simulations requiring large numbers of runs. Here an alternative suggestion (I would also change the `_benchmark_cmake` tag to `_performance_cmake`). I hope I got the formating right, it looked decent in Emacs with fixed-width font.

```suggestion
Maximize performance, reduce energy consumption
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 

The following options help to optimize NEST for maximal performance and thus reduced energy consumption.

+------------------------------------------+-----------------------------------------------+
| ``-Dwith-optimize="-O3 -march=native"``  | Activate most compiler options that do not    |
|                                          | affect compliance with IEEE754 numerics and   |
|                                          | optimize for CPU type used                    |
+------------------------------------------+-----------------------------------------------+
| ``-Dwith-defines=-DNDEBUG``              | Disable all ``assert()`` statements in NEST   |
+------------------------------------------+-----------------------------------------------+

.. note::

   * In our experience, gains from these optimizations are not very large. I can still be sensible to test them,
     especially if you are going to perform large numbers of simulations.
   * Your particular use case may touch upon corner cases in NEST execution that our extensive test suite has
     not covered. Internal consistency tests in NEST in form of ``assert()`` statements can help to detect such
     corner cases. Using the optimization options above removes these internal checks and thus increases the
     risk that NEST will produce incorrect results. Therefore, use these options *only after you have performed
     multiple simulations of your specific model with default optimization settings* (i.e., ``-O2``) and assertions
     thus in place.
   * Using ``-march=native`` requires that you build NEST on the same CPU architecture as you will use to run it.
   * For the technically minded: Even just using ``-O3`` removes some ``assert()``s from NEST since we
     have wrapped some of them in functions which get eliminated due to interprocedural optimization. ```

**Review by heplesser — CHANGES_REQUESTED** [review]
@jessica-mitchell Thanks a lot! I revised and extended the text a bit.

File: doc/htmldoc/hpc/benchmarking.rst [hunk 2]
```diff
@@ -3,22 +3,25 @@
 Benchmarking NEST
 =================
 
+.. seealso::
+
+   When compiling NEST to perform benchmarks, see :ref:`our cmake options <benchmark_cmake>` for improved performance and energy saving.
```
Comment (jessica-mitchell) [inline doc/htmldoc/hpc/benchmarking.rst:8]:
```suggestion
   When compiling NEST to perform benchmarks, see :ref:`our cmake options <performance_cmake>` for improved performance and energy saving.
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 3]
```diff
@@ -99,6 +99,28 @@ Use Python to build PyNEST
 
 For more details, see the :ref:`Python binding <compile_with_python>` section below.
 
+.. _benchmark_cmake:
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:102]:
```suggestion
.. _performance_cmake:
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 4]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+| ``-Dwith-optimize="-O3 -march=native"``  | Activate most compiler options that do not    |
+|                                          | affect compliance with IEEE754 numerics and   |
+|                                          | optimize for CPU type used                    |
++------------------------------------------+-----------------------------------------------+
+| ``-Dwith-defines=-DNDEBUG``              | Disable all ``assert()`` statements in NEST   |
++------------------------------------------+-----------------------------------------------+
+
+.. note::
+
+   * In our experience, gains from these optimizations are not very large. I can still be sensible to test them,
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:119]:
```suggestion
   * In our experience, gains from these optimizations are not very large. It can still be sensible to test them,
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 5]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+     especially if you are going to perform large numbers of simulations.
+   * Your particular use case may touch upon corner cases in NEST execution that our extensive test suite has
+     not covered. Internal consistency tests in NEST in form of ``assert()`` statements can help to detect such
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:122]:
```suggestion
     not covered. Internal consistency tests in NEST in the form of ``assert()`` statements can help to detect such
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 6]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+     corner cases. Using the optimization options above removes these internal checks and thus increases the
+     risk that NEST will produce incorrect results. Therefore, use these options *only after you have performed
+     multiple simulations of your specific model with default optimization settings* (i.e., ``-O2``) and assertions
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:125]:
```suggestion
     multiple simulations of your specific model with default optimization settings* (i.e., ``-O2``), which leaves the assertions
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 7]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+     thus in place.
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:126]:
```suggestion
     in place.
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 8]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+   * Using ``-march=native`` requires that you build NEST on the same CPU architecture as you will use to run it.
+   * For the technically minded: Even just using ``-O3`` removes some ``assert()``s from NEST since we
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:128]:
```suggestion
   * For the technically minded: Even just using ``-O3`` removes some ``assert()`` calls from NEST since we
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 9]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+     have wrapped some of them in functions which get eliminated due to interprocedural optimization. ```
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:129]:
```suggestion
     have wrapped some of them in functions, which get eliminated due to interprocedural optimization. 
```

File: doc/htmldoc/installation/cmake_options.rst [hunk 10]
```diff
@@ -99,6 +99,37 @@ Use Python to build PyNEST
...
+|                                          | optimize for CPU type used                    |
++------------------------------------------+-----------------------------------------------+
+| ``-Dwith-defines=-DNDEBUG``              | Disable all ``assert()`` statements in NEST   |
++------------------------------------------+-----------------------------------------------+
+
+.. note::
+
+   * In our experience, gains from these optimizations are not very large. I can still be sensible to test them,
+     especially if you are going to perform large numbers of simulations.
+   * Your particular use case may touch upon corner cases in NEST execution that our extensive test suite has
```
Comment (jessica-mitchell) [inline doc/htmldoc/installation/cmake_options.rst:121]:
```suggestion
   * Your particular use case may contain edge cases during NEST execution that our extensive test suite has
```

**Review by heplesser — APPROVED** [review]

**Comment by jessica-mitchell** [comment]
@heplesser can we merge this with one reviewer, or do you want others to comment?

**Comment by heplesser** [comment]
> @heplesser can we merge this with one reviewer, or do you want others to comment?

Let's wait a little :).

**Review by ackurth — APPROVED** [review]
This is a very clear and understandable descriptions.
Looks good to me!
